### PR TITLE
Fixing google play variant push notifications being skipped

### LINKED ISSUE #5038: Missing push notifications (Google Play variant)
Centralising #5028 #5016 #4862 #4393 

We're getting reports that the google play variant of the app is missing/skipping notifications, this also manifests as notifications not being marked as read when reading from other sessions

To help troubleshoot this

- What date and time did the sender send the missed message (with timezone information) - We'll cross reference with push server outages
- Are you using multiple clients, if so, have any of the notification settings been modified?
- Does the issue happen for all rooms/dms?
- Are you running a private homeserver, if so do you know the version? `{homeserver_url}/_synapse/admin/v1/server_version` eg https://matrix-client.matrix.org/_synapse/admin/v1/server_version

---
 
- Whilst having the issue, screenshot the list in Settings -> Advanced settings -> Push rules. We're looking to see if a rule has incorrectly started filtering out messages
- Does Element receive raw pushes? A raw `Push` counter can be displayed with  Advanced settings -> Enable `Developer mode` -> Enable `Show debug info on screen`.  

---

- Does switching room notifications to `none` and back to `all messages` fix the issue?

| PUSH RULES - 1| PUSH RULES - 2 |
| --- | --- |
Screenshot_20220124_171938|Screenshot_20220124_171945

| RAW PUSH COUNTER | ELEMENT WEB NOTIFICATION SETTINGS | ELEMENT WEB ROOM SETTINGS | 
| --- | --- | --- |
2022-01-24T12:01:50,573509360+00:00|2022-01-24T16:21:50,592516708+00:00|2022-01-24T16:23:21,958254940+00:00

### PR DESCRIPTION
**Pull request by ouchadam**
Fixes #5038 Missing google play variant push notifications

**tl;dr** The _not so ideal workaround_ is to disable wifi, as we only run the fast lane pipeline whilst the device is on a wifi connection.

The bug -

Push messages received whilst your own message is the latest in a room whilst a background sync is occurring are skipped until the background sync completes.

The conditions to encounter this bug are:

- a message from yourself is the latest local message in the room
- the device is on wifi
- the app is in the background
- receive a push notification from FCM (google play variant only)
- the fast lane `getEvent` responds faster than the scheduled `/sync`  
- any further pushes received whilst the `/sync` is running are skipped
- after the sync completes the latest message in the room is no longer yourself, breaking the cycle

The longer the sync takes, the more potential messages are skipped and each time your own message is the last in the room the bug comes back.

Further details - 

The issue is caused by us checking if the newly received message (event) has already been read and incorrectly inferring that it has been read when it may not have even been seen yet! 

Our _is already read_ rules were inferring that because the latest locally known event was from ourselves that the new event must have already been read, this is problematic because when an event comes from a FCM push it goes through a _fast lane_ pipeline which doesn't insert the new event, only the `/sync` pipeline inserts the event.

This means that whilst a push triggered `/sync` is running any pushes (including the original trigger) that goes through the fast lane pipeline end up being marked as read which in turn skips them from being displayed as notifications.

When the sync finishes and inserts the new message events the latest event in the room is no longer ourselves and future notifications will work as expected.

-----

The fix is to only check the latest room event if the new event doesn't exist

- reworks the `isEventRead` logic to always check if the new event exists locally and if not, inferring that it must not have been read
- took the opportunity to flatten all of the execution branches

*with a hardcoded delay of 5 seconds for the sync

| BEFORE | AFTER |
| --- | --- |
|before-missed-notifications|after-missing-notification-2

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt [hunk 1]
```diff
@@ -34,27 +34,29 @@ internal fun isEventRead(realmConfiguration: RealmConfiguration,
     if (LocalEcho.isLocalEchoId(eventId)) {
         return true
     }
-    // If we don't know if the event has been read, we assume it's not
-    var isEventRead = false
 
-    Realm.getInstance(realmConfiguration).use { realm ->
-        val latestEvent = TimelineEventEntity.latestEvent(realm, roomId, true)
-        // If latest event is from you we are sure the event is read
-        if (latestEvent?.root?.sender == userId) {
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/database/query/ReadQueries.kt:43]: moving this check to be after `eventToCheck == null` is the fix

**Review by bmarty — APPROVED** [review]
Thanks for your investigation on this issue.
The fix looks sensible.

**Comment by DoM1niC** [comment]
seems to work fine here
